sap.ui.core.CSSSize supports css calc()

themasch opened the pull request:
First things first: **I hereby declare to agree to the OpenUI5 Individual Contributor License Agreement.**

This PR should fix #185. 

Should I add more test cases? 

There is one little problem with the current implementation (beside the regexp beeing way to big): It doesn't check for logical errors like "3 + 10rem". 
Additionally it does not support features like attr() in calc() (yet).

I don't expect this PR to be pulled immediately, just wanted to get some input on how to get this done.

themasch commented:
Whats the current status on this? We are already using this "fix" in some prototypes. 

Aside: How are these Controls which are calculation using CSSSize when the value ist "auto" or "inherit", because these values are allowed in the current RegEx ;)

akudev commented:
Hi,
sorry for the long delay... the discussion is not through yet (and apparently not at the top of our priority list right now...). 
@codeworrior also says that the RegExp needs to be changed because the ignore-case modifier (the "/i") may not be used because it breaks the transformation to XSD schema. (yeah, you cannot know this, but it's like that)

Regards
Andreas

themasch commented:
Hi. I just updated the RegExp to reflect changes between 1.24 and 1.28 and also refactored it to be more human read- and maintainable. The old one got really really long.. 
QUnit tests are green, performance will be lower but not terrible. 

Does the new format "breaks the transformation to XSD schema", too?

codeworrior commented:
I'm sorry, but this refactoring totally breaks the static code analysis that we are doing to restore the metamodel description of the types and which is necessary to create the XSD from (as well as other stuff).

Please keep the RegExp as a single literal with no modifiers and with explicit anchors ^(...)$ (and no other anchors inside). 

Sorry for this inconvenience, but that's how it works currently. 

We are currently composing some documentation about these constraints (such documentation is long overdue, I have to admit). With that, the constraints hopefully become more transparent (but still not more convenient).

Regards, Frank

themasch commented:
like this?

codeworrior commented:
Yes. After taking a short look, I would say the syntax can be understood by our code analysis and also should be convertible to an XSD RegExp (we cut off the anchors ^(...)$).

codeworrior commented:
Status update:

I took your pull request and created a change for internal review from it. 

While doing so, I noticed a few minor shortcomings in the regular expression. I hope you don't mind that I modified the regexp to bring it closer to the spec (http://dev.w3.org/csswg/css-values-3/#calc-syntax):

 - make it more strict wrt. the necessary whitespace around  the '-' and '+' operators
 - allow optional whitespace everywhere in an expression
 - add (simple) support for parenthesis to allow grouping of operations
 - make the operator mandatory if there is more than one value 

As we are currently limited to a regexp to define the syntax limitations, we can't fully check the input against the above spec. Type mismatches (e.g. adding px + em) or unbalanced parenthesis will go unnoticed. Nevertheless, my minimal goal would be that any valid value will pass the regexp and as many invalid values as possible are rejected by it. 

Hope you're fine with that.

themasch commented:
Sure, thats great.
I noticed the whitespace problem already and fixed it somewhere internally 
but forgot to update that PR.

Nice to see that this things are making progress with this PR. Thanks!

Gesendet mit AquaMail für Android
http://www.aqua-mail.com

codeworrior commented:
The pull request now has been merged. The enhanced CSSSize type will be part of the upcoming 1.30 version of OpenUI5. 

Thanks again for this contribution!
